File: cpp/src/networks/mlp.cpp
#include <rzsm/networks/mlp.hpp>

namespace rzsm {

torch::nn::Sequential build_mlp(
    const std::vector<int64_t>& sizes,
    ActivationType activation,
    ActivationType output_activation)
{
    torch::nn::Sequential seq;
    for (size_t j = 0; j + 1 < sizes.size(); ++j) {
        seq->push_back(torch::nn::Linear(sizes[j], sizes[j + 1]));
        ActivationType act = (j < sizes.size() - 2) ? activation : output_activation;
        push_activation(seq, act);
    }
    return seq;
}

// ── MLPActor ─────────────────────────────────────────────────────

MLPActorImpl::MLPActorImpl(
    int64_t obs_dim, int64_t act_dim,
    const std::vector<int64_t>& hidden_sizes,
    ActivationType activation, double act_limit)
    : obs_dim_(obs_dim), act_dim_(act_dim),
      hidden_sizes_(hidden_sizes), activation_(activation),
      act_limit_(act_limit)
{
    std::vector<int64_t> sizes = {obs_dim};
    sizes.insert(sizes.end(), hidden_sizes.begin(), hidden_sizes.end());
    sizes.push_back(act_dim);
    pi_ = build_mlp(sizes, activation, ActivationType::Tanh);
    register_module("pi", pi_);
}

torch::Tensor MLPActorImpl::forward(torch::Tensor obs) {
    return act_limit_ * pi_->forward(obs);
}

// ── MLPQFunction ─────────────────────────────────────────────────

MLPQFunctionImpl::MLPQFunctionImpl(
    int64_t obs_dim, int64_t act_dim,
    const std::vector<int64_t>& hidden_sizes,
    ActivationType activation)
    : obs_dim_(obs_dim), act_dim_(act_dim),
      hidden_sizes_(hidden_sizes), activation_(activation)
{
    std::vector<int64_t> sizes = {obs_dim + act_dim};
    sizes.insert(sizes.end(), hidden_sizes.begin(), hidden_sizes.end());
    sizes.push_back(1);
    q_ = build_mlp(sizes, activation);
    register_module("q", q_);
}

torch::Tensor MLPQFunctionImpl::forward(torch::Tensor obs, torch::Tensor act) {
    return q_->forward(torch::cat({obs, act}, /*dim=*/-1)).squeeze(-1);
}

// ── AdversarialMLPQFunction ──────────────────────────────────────

AdversarialMLPQFunctionImpl::AdversarialMLPQFunctionImpl(
    int64_t obs_dim, int64_t act_dim, int64_t dist_dim,
    const std::vector<int64_t>& hidden_sizes,
    ActivationType activation)
    : obs_dim_(obs_dim), act_dim_(act_dim), dist_dim_(dist_dim),
      hidden_sizes_(hidden_sizes), activation_(activation)
{
    std::vector<int64_t> sizes = {obs_dim + act_dim + dist_dim};
    sizes.insert(sizes.end(), hidden_sizes.begin(), hidden_sizes.end());
    sizes.push_back(1);
    q_ = build_mlp(sizes, activation);
    register_module("q", q_);
}

torch::Tensor AdversarialMLPQFunctionImpl::forward(
    torch::Tensor obs, torch::Tensor act, torch::Tensor act_d)
{
    return q_->forward(torch::cat({obs, act, act_d}, /*dim=*/-1)).squeeze(-1);
}

} // namespace rzsm

File: src/networks/mlp.py
"""MLP network building blocks for TD3 actor-critic architectures."""

from typing import List, Optional, Type

import numpy as np
import torch
import torch.nn as nn


def mlp(
    sizes: List[int],
    activation: Type[nn.Module],
    output_activation: Optional[Type[nn.Module]] = None,
) -> nn.Sequential:
    """Build a feedforward MLP from a list of layer sizes."""
    if output_activation is None:
        output_activation = nn.Identity
    layers: list[nn.Module] = []
    for j in range(len(sizes) - 1):
        act = activation if j < len(sizes) - 2 else output_activation
        layers += [nn.Linear(sizes[j], sizes[j + 1]), act()]
    return nn.Sequential(*layers)


def count_vars(module: nn.Module) -> int:
    return sum(np.prod(p.shape) for p in module.parameters())


class MLPActor(nn.Module):
    """Deterministic policy: obs -> tanh-squashed action scaled by act_limit."""

    def __init__(
        self,
        obs_dim: int,
        act_dim: int,
        hidden_sizes: tuple[int, ...],
        activation: Type[nn.Module],
        act_limit: float,
    ):
        super().__init__()
        pi_sizes = [obs_dim] + list(hidden_sizes) + [act_dim]
        self.pi = mlp(pi_sizes, activation, output_activation=nn.Tanh)
        self.act_limit = act_limit

    def forward(self, obs: torch.Tensor) -> torch.Tensor:
        return self.act_limit * self.pi(obs)


class MLPQFunction(nn.Module):
    """Q-function for single-agent: Q(obs, act) -> scalar."""

    def __init__(
        self,
        obs_dim: int,
        act_dim: int,
        hidden_sizes: tuple[int, ...],
        activation: Type[nn.Module],
    ):
        super().__init__()
        self.q = mlp([obs_dim + act_dim] + list(hidden_sizes) + [1], activation)

    def forward(self, obs: torch.Tensor, act: torch.Tensor) -> torch.Tensor:
        return torch.squeeze(self.q(torch.cat([obs, act], dim=-1)), -1)


class AdversarialMLPQFunction(nn.Module):
    """Q-function for two-player game: Q(obs, act_ctrl, act_dist) -> scalar."""

    def __init__(
        self,
        obs_dim: int,
        act_dim: int,
        dist_dim: int,
        hidden_sizes: tuple[int, ...],
        activation: Type[nn.Module],
    ):
        super().__init__()
        self.q = mlp(
            [obs_dim + act_dim + dist_dim] + list(hidden_sizes) + [1], activation
        )

    def forward(
        self, obs: torch.Tensor, act: torch.Tensor, act_d: torch.Tensor
    ) -> torch.Tensor:
        return torch.squeeze(self.q(torch.cat([obs, act, act_d], dim=-1)), -1)
